src/plugins/graph/models/graph-model.ts

```diff
@@ -168,20 +168,87 @@ export const GraphModel = TileContentModel
     },
     setShowMeasuresForSelection(show: boolean) {
       self.showMeasuresForSelection = show;
+    },
+  }))
+  .actions(self => ({
+    linkDataSet() {
+      if (!self.data) {
+        console.warn("GraphModel.linkDataSet requires a dataset");
+        return;
+      }
+      self.config.setDataset(self.data);
+      self.setAttributeID("x", self.data.attributes[0].id);
+      self.setAttributeID("y", self.data.attributes[1].id);
+    },
+    unlinkDataSet() {
+      if (self.data) {
+        console.warn("GraphModel. unlinkDataSet expects the dataset to be unlinked");
+        return;
+      }
+      self.setAttributeID("y", "");
+      self.setAttributeID("x", "");
+      self.config.setDataset(undefined);
     }
   }))
   .actions(self => ({
-    updateAfterSharedModelChanges(sharedModel?: SharedModelType, changeType?: SharedModelChangeType) {
-      if (changeType === "link" && self.data) {
+    updateAfterSharedModelChanges(sharedModel?: SharedModelType) {
+      // We need to figure out how to know if we need to update the
+      // dataSet. The config.dataSet is volatile, but setting it
+      // might also update state in the config I'm not sure
+      // We could just check if they match and then update it if
+      // not. And then we'd also need a reaction that does the same
+      // thing, but we'd need the reaction to only do this if it isn't
+      // being done here.
+
+
+      // Note this will also happen in the reaction below
+      // we do it here just to be safe incase this function is called
+      // first
+      if (self.data !== self.config.dataset) {
         self.config.setDataset(self.data);
-        self.setAttributeID("x", self.data.attributes[0].id);
-        self.setAttributeID("y", self.data.attributes[1].id);
       }
-      else if (changeType === "unlink") {
+
+      if (!self.data) {
+        // When we have no dataset default the graph to x y axes
         self.setAttributeID("y", "");
         self.setAttributeID("x", "");
-        self.config.setDataset(undefined);
+      } else {
+        // If our graph doesn't have useful axes then set them
+        // This will typically happen when the graph is linked
+        if (!self.getAttributeID("y") && !self.getAttributeID("x")) {
+          self.setAttributeID("x", self.data.attributes[0].id);
+          self.setAttributeID("y", self.data.attributes[1].id);
+        }
       }
+    },
+    afterAttach() {
+      addDisposer(self, reaction(
+        () => self.data,
+        (data, prevData, more) => {
+          // CHECKME: this will only work correctly if setDataset doesn't
+          // trigger any state updates
```

Comment:
Has this been checked? If it's true, it might be worth putting a comment above `setDataset` saying it can't trigger any state updates to make sure programmers of the future don't make any grave mistakes.